pkg/replicater/scheme.go

```diff
@@ -0,0 +1,387 @@
+		targetMetaFileContent, err := ioutil.ReadAll(targetMetaFile)
+		if err != nil {
+			return fmt.Errorf("read target meta file: %w", err)
+		}
+
+		if bytes.Equal(originMetaFileContent, targetMetaFileContent) {
+			// If the origin meta file content and target meta file content is
+			// equal, we know we have already successfully replicated
+			// previously.
+			level.Debug(rs.logger).Log("msg", "skipping block as already replicated", "block_uuid", id.String())
+			rs.metrics.blocksAlreadyReplicated.Inc()
+
+			return nil
+		}
+	}
+
+	if err := rs.fromBkt.Iter(ctx, chunksDir, func(objectName string) error {
+		err := rs.ensureObjectReplicated(ctx, objectName)
+		if err != nil {
+			return fmt.Errorf("replicate object %v: %w", objectName, err)
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if err := rs.ensureObjectReplicated(ctx, indexFile); err != nil {
+		return fmt.Errorf("replicate index file: %w", err)
+	}
+
+	level.Debug(rs.logger).Log("msg", "replicating meta file", "object", metaFile)
+
+	if err := rs.toBkt.Upload(ctx, metaFile, bytes.NewReader(originMetaFileContent)); err != nil {
+		return fmt.Errorf("upload meta file: %w", err)
+	}
+
+	rs.metrics.blocksReplicated.Inc()
+
+	return nil
+}
+
+// ensureBlockIsReplicated ensures that an object present in the origin bucket
+// is present in the target bucket.
+func (rs *replicationScheme) ensureObjectReplicated(ctx context.Context, objectName string) error {
+	level.Debug(rs.logger).Log("msg", "ensuring object is replicated", "object", objectName)
+
+	exists, err := rs.toBkt.Exists(ctx, objectName)
+	if err != nil {
+		return fmt.Errorf("check if %v exists in target bucket: %w", objectName, err)
+	}
+
+	// skip if already exists.
+	if exists {
+		level.Debug(rs.logger).Log("msg", "skipping object as already replicated", "object", objectName)
+		return nil
+	}
+
+	level.Debug(rs.logger).Log("msg", "object not present in target bucket, replicating", "object", objectName)
+
+	r, err := rs.fromBkt.Get(ctx, objectName)
+	if err != nil {
+		return fmt.Errorf("get %v from origin bucket: %w", objectName, err)
+	}
+
+	defer r.Close()
+
+	if err = rs.toBkt.Upload(ctx, objectName, r); err != nil {
+		return fmt.Errorf("upload %v to target bucket: %w", objectName, err)
+	}
+
+	level.Info(rs.logger).Log("msg", "object replicated", "object", objectName)
+	rs.metrics.objectsReplicated.Inc()
+
+	return nil
+}
+
+// loadMeta loads the meta.json from the origin bucket and returns the meta
+// struct as well as if failed, whether the failure was due to the meta.json
+// not being present or partial. The distinction is important, as if missing or
+// partial, this is just a temporary failure, as the block is still being
+// uploaded to the origin bucket.
+func loadMeta(ctx context.Context, bucket objstore.BucketReader, id ulid.ULID) (*metadata.Meta, bool, error) {
+	src := path.Join(id.String(), thanosblock.MetaFilename)
+
+	r, err := bucket.Get(ctx, src)
+	if bucket.IsObjNotFoundErr(err) {
```

Comment:
Don't we forget to do `r.Close()` here if the object has not been found?

Reply:
Good catch!

Reply:
Good catch, thanks!

Reply:
@GiedriusS refer to [test](https://circleci.com/gh/thanos-io/thanos/7171?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link), if close here, it would raise a error, the older one is fine i think, so revert change.

Reply:
You are correct, it seems like there is some kind of implicit invariant here in our object storage implementations that no `io.ReadCloser` is returned if the object doesn't exist.